framework/src/org/checkerframework/framework/type/AnnotationClassLoader.java

```diff
@@ -683,43 +676,52 @@ public int compare(File o1, File o2) {
      *
      * @param fullyQualifiedClassName the fully qualified name of the class
-     * @return the loaded annotation class if it is defined with ElementType.TYPE_USE and is a
-     *     supported annotation, null otherwise
+     * @return a pair consisting of a boolean indicating whether the class could be loaded, and the
+     *     loaded annotation class if it has a {@code @Target} meta-annotation with the required
+     *     ElementType values, and is a supported annotation by a checker. If the class cannot be
+     *     loaded, the pair {@code <false, null>} is returned. If an annotation class was loaded but
+     *     is not supported by a checker, the pair {@code <true, null>} is returned. If the
+     *     annotation is loaded and supported, then the pair {@code <true, theLoadedClass>} is
+     *     returned.
      */
-    private final /*@Nullable*/ Class<? extends Annotation> loadAnnotationClass(
+    protected final Pair<Boolean, /* @Nullable */ Class<? extends Annotation>> loadAnnotationClass(
             final String fullyQualifiedClassName) {
         Class<?> cls = null;
 
         try {
             // load the class
-            cls = Class.forName(fullyQualifiedClassName, true, getAppClassLoader());
+            ClassLoader classLoader = getClassLoader();
+            if (classLoader != null) {
+                cls = Class.forName(fullyQualifiedClassName, true, classLoader);
+            }
         } catch (ClassNotFoundException e) {
             // do nothing: projects can have annotation class files and regular
             // source files located within the same directory, and as such when
             // it tires to load an uncompiled source file, it will throw
             // ClassNotFoundException
         }
 
-        // ensure that the freshly loaded class is an annotation, and has
-        // the @Target annotation
-        if (cls != null && cls.isAnnotation() && cls.getAnnotation(Target.class) != null) {
-            // retrieve the set of ElementTypes in the @Target
-            // meta-annotation and check to see if this annotation is
-            // supported for automatic loading
-            if (AnnotatedTypes.hasTypeQualifierElementTypes(
-                    cls.getAnnotation(Target.class).value(), cls)) {
-                // if it is supported, then subclass it as an Annotation
-                // class
-                Class<? extends Annotation> annoClass = cls.asSubclass(Annotation.class);
-
-                // see if the annotation is supported by a checker
+        // ensure that the freshly loaded class is an annotation
+        if (cls != null && cls.isAnnotation()) {
+            Class<? extends Annotation> annoClass = cls.asSubclass(Annotation.class);
+
+            // make sure the loaded annotation class has the @Target meta-annotation, and is a well
+            // defined qualifier or is the special qualifier @Unqualified
+            if (annoClass.getAnnotation(Target.class) != null
```

Comment:
Extract to helper method, to filter
see if we already have this logic somewhere